lib/Crypto/doc/sailfish-crypto-overview.qdoc

```diff
@@ -0,0 +1,218 @@
+no secret key data) will be returned to the client application.
+
+Note that these operations are all asynchronous, however in the snippet we
+force the operation to be synchronous by calling the \tt{waitForFinished()}
+method on the reply object.  In practice, the client application should instead
+use a watcher object which will notify when the operation is complete.
+
+\code
+// Set the key template metadata.
+Sailfish::Crypto::Key keyTemplate, symmetricKeyReference;
+keyTemplate.setAlgorithm(Sailfish::Crypto::Key::Aes256);
+keyTemplate.setOrigin(Sailfish::Crypto::Key::OriginDevice);
+keyTemplate.setBlockModes(Sailfish::Crypto::Key::BlockModeCBC);
+keyTemplate.setEncryptionPaddings(Sailfish::Crypto::Key::EncryptionPaddingNone);
+keyTemplate.setSignaturePaddings(Sailfish::Crypto::Key::SignaturePaddingNone);
+keyTemplate.setDigests(Sailfish::Crypto::Key::DigestSha256);
+keyTemplate.setOperations(Sailfish::Crypto::Key::Encrypt
+                         |Sailfish::Crypto::Key::Decrypt);
+
+// Set the identifier for the key.
+// This assumes the existence of an "ExampleCollection" secure storage
+// collection, in which the key will be stored.
+// See Sailfish::Secrets::SecretManager::createCollection().
+keyTemplate.setIdentifier(QStringLiteral("ExampleKey"),
+                          QStringLiteral("ExampleCollection"));
+
+// Ask the system service to generate and store the key securely.
+Sailfish::Crypto::CryptoManager cm;
+QDBusPendingReply<Sailfish::Crypto::Result, Sailfish::Crypto::Key>
+        generateReply = cm.generateStoredKey(
+                keyTemplate,
+                Sailfish::Crypto::CryptoManager::DefaultCryptoStoragePluginName);
+generateReply.waitForFinished();
+if (generateReply.argumentAt<0>().code() == Sailfish::Crypto::Result::Failed) {
+    qWarning() << "Unable to generate and store symmetric key:"
+               << generateReply.argumentAt<0>().errorMessage();
+} else {
+    symmetricKeyReference = generateReply.argumentAt<1>();
+}
+\endcode
+
+\section4 Encrypting data with a symmetric key
+
+After generating a symmetric key, that key may be used to encrypt data.
+Note that the key may be either a key reference (that is, a key which contains
+only metadata and a valid identifier, which references a full key stored in
+secure storage) or a full key (that is, a key which contains secret key data).
+
+In this example, we use the reference key which was returned to the application
+in the previous snippet.
+
+\code
+QByteArray ciphertext, plaintext = "Example plaintext data";
+QDBusPendingReply<Sailfish::Crypto::Result, QByteArray>
+        encryptReply = cm.encrypt(
+                plaintext,
+                symmetricKeyReference,
+                Sailfish::Crypto::Key::BlockModeCBC,
+                Sailfish::Crypto::Key::EncryptionPaddingNone,
+                Sailfish::Crypto::Key::DigestSha256,
+                Sailfish::Crypto::CryptoManager::DefaultCryptoStoragePluginName);
```

Comment:
The style here is inconsistent with the above key template example, like discussed earlier Qt-style tries to avoid functions with a lot of parameters as it is not easy remember the order of passing (and in case of integers and strings there is a risk wrong parameters get passed without compiler complaining).

Reply:
iOS provides a map where you must put all this [information](https://www.andyibanez.com/using-ios-keychain/):

```objective-c
    //Let's create an empty mutable dictionary:
    NSMutableDictionary *keychainItem = [NSMutableDictionary dictionary];
     
    NSString *username = self.nwUsernameTxtFld.text;
    NSString *password = self.passwordTxtFld.text;
    NSString *website = @"<a class="vglnk" href="http://www.myawesomeservice.com" rel="nofollow"><span>http</span><span>://</span><span>www</span><span>.</span><span>myawesomeservice</span><span>.</span><span>com</span></a>";
     
    //Populate it with the data and the attributes we want to use.
     
    keychainItem[(__bridge id)kSecClass] = (__bridge id)kSecClassInternetPassword; // We specify what kind of keychain item this is.
    keychainItem[(__bridge id)kSecAttrAccessible] = (__bridge id)kSecAttrAccessibleWhenUnlocked; // This item can only be accessed when the user unlocks the device.
    keychainItem[(__bridge id)kSecAttrServer] = website;
    keychainItem[(__bridge id)kSecAttrAccount] = username;
```

Here we will never have a mistake. We could also have a struct for these purposes.

The actual processing of this data is performed by only one function call with only one argument - this dictionary. It returns boolean result. Simple as possible. We may simply look what other systems provide and do something similar.

Reply:
Yes, I will do something similar for the client API when I create the wrapper, with per-request-type request objects.

Reply:
Nice!